### PR TITLE
Extract view hierarchy description attachment code into attachment generator

### PR DESCRIPTION
**Pull request by NickEntin**
This is a step towards pulling functionality out of `ARKEmailBugReporter` and into small utility classes, in order to make it easier to write a custom bug reporter.

### CODE REVIEW

**Comment by NickEntin** [comment]
🤦 I just merged #103 into this branch

**Comment by NickEntin** [comment]
Okay, removed the changes from #103. This should be good to go again.

**Review by efirestone — APPROVED** [review]

File: Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+@objc(ARKViewHierarchyAttachmentGenerator)
+public final class ViewHierarchyAttachmentGenerator: NSObject {
+
+    // MARK: - Public Static Methods
+
+    /// Captures a textual representation of the current view hierarchy and returns a bug report attachment containing
+    /// plain text data for that representation.
+    @objc
+    public static func captureCurrentHierarchy() -> ARKBugReportAttachment {
+        return generateAttachment(for: UIApplication.shared.windows)
```
Comment (dfed) [inline Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift:28]: nit: you don't need this `return` 😉 
```suggestion
        generateAttachment(for: UIApplication.shared.windows)
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+@testable import Aardvark
+
+final class ViewHierarchyAttachmentGeneratorTests: XCTestCase {
+
+    func testAttachmentMetadata() {
+        let attachment = ViewHierarchyAttachmentGenerator.generateAttachment(for: [])
+
+        XCTAssertEqual(attachment.fileName, "view_hierarchy.txt")
+        XCTAssertEqual(attachment.dataMIMEType, "text/plain")
+    }
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:28]: It's hard to look at this method name and understand what will be tested. I think it'd be best to bring these two `XCTAssert` calls into two separate tests. I'd write this as:

```suggestion
    func test_generateAttachment_hasExpectedFilename() {
        let attachment = ViewHierarchyAttachmentGenerator.generateAttachment(for: [])

        XCTAssertEqual(attachment.fileName, "view_hierarchy.txt")
    }

    func test_generateAttachment_hasExpectedMIMEType() {
        let attachment = ViewHierarchyAttachmentGenerator.generateAttachment(for: [])

        XCTAssertEqual(attachment.dataMIMEType, "text/plain
    }
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    func testDescriptionForSimpleWindow() {
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:30]: Following the naming convention above:
```suggestion
    func test_generateAttachment_description_hasExpectedFormatForSimpleWindow() {
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        XCTAssertEqual(
+            description,
+            """
+            Window Description
+
+            """
+        )
+    }
+
+    func testDescriptionForWindowWithSubviews() {
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:44]: Following the naming convention above:
```suggestion
    func test_generateAttachment_description_hasExpectedFormatForWindowWithSubviews() {
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            Window Description
+               | View Description for B
+               |    | View Description for C
+               | View Description for A
+
+            """
+        )
+    }
+
+    func testDescriptionForWindowWithSublayers() {
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:68]: Following the naming convention above:
```suggestion
    func test_generateAttachment_description_hasExpectedFormatForWindowWithSublayers() {
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            Window Description
+               | View Description for Subview
+               |    | Layer Description
+               |    | Layer Description
+
+            """
+        )
+    }
+
+    func testDescriptionForWindowWithViewController() {
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:92]: Following the naming convention above:
```suggestion
    func test_generateAttachment_description_hasExpectedFormatForWindowWithViewController() {
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+}
+
+// MARK: -
+
+private final class TestWindow: UIWindow {
+
+    override var description: String {
+        return "Window Description"
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:120]: nit, no need for the `return`!
```suggestion
        "Window Description"
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    // MARK: - Private Properties
+
+    private let identifier: String
+
+    // MARK: - UIView
+
+    override var description: String {
+        return "View Description for \(identifier)"
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:147]:
```suggestion
        "View Description for \(identifier)"
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+}
+
+private final class TestLayer: CALayer {
+
+    override var description: String {
+        return "Layer Description"
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:155]:
```suggestion
        "Layer Description"
```

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+private final class TestViewController: UIViewController {
+
+    override func loadView() {
+        view = TestView(identifier: "Managed View")
+    }
+
+    override var description: String {
+        return "View Controller Description"
```
Comment (dfed) [inline Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift:167]:
```suggestion
        "View Controller Description"
```

**Review by dfed — APPROVED** [review]
LGTM with some non-blocking thoughts/nits

File: Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift (hunk 1 above)
Comment (NickEntin) [reply Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift:28]: Xcode 10 support 😉

Comment (dfed) [reply Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift:28]: giphy

... probably time to drop that? but can be a separate PR.

Comment (NickEntin) [reply Sources/Aardvark/Bug Reporting/ViewHierarchyAttachmentGenerator.swift:28]: Oh wait, actually I guess I already bumped to Xcode 11 🤔

File: Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift (hunk 2 above)
Comment (NickEntin) [reply Sources/AardvarkTests/ViewHierarchyAttachmentGeneratorTests.swift]: Addressed in f3b8f0a.

**Pull request merged**